packages/@expo/cli/src/start/server/metro/createExpoFallbackResolver.ts

```diff
@@ -0,0 +1,160 @@
+import fs from 'fs';
+import type { ResolutionContext } from 'metro-resolver';
+import path from 'path';
+
+import type { StrictResolver, StrictResolverFactory } from './withMetroMultiPlatform';
+import type { ExpoCustomMetroResolver } from './withMetroResolvers';
+
+/** A record of dependencies that we know are only used for scripts and config-plugins */
+const EXCLUDE_ORIGIN_MODULES: Record<string, true | undefined> = {
+  '@expo/config': true,
+  '@expo/config-plugins': true,
+  'schema-utils': true, // Used by `expo-router/plugin`
+  semver: true, // Used by `expo-router/doctor`
+};
+
+interface PackageMetaPeerDependenciesMetaEntry {
+  [propName: string]: unknown;
+  optional?: boolean;
+}
+
+interface PackageMeta {
+  [propName: string]: unknown;
+  dependencies?: Record<string, unknown>;
+  peerDependencies?: Record<string, unknown>;
+  peerDependenciesMeta?: Record<string, PackageMetaPeerDependenciesMetaEntry | undefined | null>;
+}
+
+interface ModuleDescription {
+  originModulePath: string;
+  moduleTestRe: RegExp;
+}
+
+const debug = require('debug')('expo:start:server:metro:fallback-resolver') as typeof console.log;
+
+/** Converts a list of module names to a regex that may either match bare module names or sub-modules of modules */
+const dependenciesToRegex = (dependencies: string[]) =>
+  new RegExp(`^(?:${dependencies.join('|')})(?:$|/)`);
+
+/** Resolves an origin module and outputs a filter regex and target path for it */
+const getModuleDescriptionWithResolver = (
+  resolve: StrictResolver,
+  originModuleName: string
+): ModuleDescription | null => {
+  let filePath: string | undefined;
+  let packageMeta: PackageMeta;
+  try {
+    const resolution = resolve(`${originModuleName}/package.json`);
```

Comment:
There are some built-in methods on the Resolution context that may resolve better/faster `context.getPackage(pkgFile)`

Reply:
done!